src/executioners/SpectralExecutionerLinearElastic.C

```diff
@@ -0,0 +1,402 @@
+        Real nu0 = _young_modulus * _average_factor / (2 * (1 + _poisson_ratio));
+        Real constant = (lambda0 + nu0) / (nu0 * (lambda0 + 2.0 * nu0));
+
+        for (int i = 0; i < ndim; i++)
+          for (int j = 0; j < ndim; j++)
+            for (int k = 0; k < ndim; k++)
+              for (int l = 0; l < ndim; l++)
+              {
+                Complex q_square = freq[0] * freq[0] + freq[1] * freq[1] + freq[2] * freq[2];
+                if (std::abs(q_square) > 1.0e-12)
+                {
+                  gamma_hat.reciprocalSpace()[index](i, j, k, l) =
+                      -1.0 * constant * (freq[i] * freq[j] * freq[k] * freq[l]) /
+                          (q_square * q_square) +
+                      (deltaij(i, k) * freq[j] * freq[l] + deltaij(j, k) * freq[i] * freq[l] +
+                       deltaij(i, l) * freq[j] * freq[k] + deltaij(j, l) * freq[i] * freq[k]) /
+                          (4.0 * nu0 * q_square);
+                }
+                else
+                {
+                  gamma_hat.reciprocalSpace()[index] = elasticity_tensor.invSymm();
+                }
+              }
+
+        index++;
+      }
+}
+
+FFTBufferBase<RankTwoTensor> &
+SpectralExecutionerLinearElastic::getInitialStress(
+    FFTBufferBase<RankTwoTensor> & epsilon_buffer,
+    const FFTBufferBase<RankFourTensor> & elastic_tensor)
+{
+  auto & stress_buffer = getFFTBuffer<RankTwoTensor>("stress");
+  const auto & grid = epsilon_buffer.grid();
+  int ni = grid[0];
+  int nj = grid[1];
+  int nk = grid[2];
+
+  // Homogeneous initial state of strain
+  fillOutEpsilonBuffer(epsilon_buffer);
+
+  size_t index = 0;
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        stress_buffer.realSpace()[index] =
+            elastic_tensor.realSpace()[index] * epsilon_buffer.realSpace()[index];
+        index++;
+      }
+
+  return stress_buffer;
+}
+
+void
+SpectralExecutionerLinearElastic::advanceReciprocalEpsilon(
+    FFTBufferBase<RankTwoTensor> & epsilon_buffer,
+    const FFTBufferBase<RankTwoTensor> & stress_buffer,
+    const FFTBufferBase<RankFourTensor> & gamma_hat)
+{
+  const auto & grid = epsilon_buffer.grid();
+  int ni = grid[0];
+  int nj = grid[1];
+  int nk = (grid[2] >> 1) + 1;
+
+  size_t index = 0;
+
+  const auto & ivec = gamma_hat.kTable(0);
+  const auto & jvec = gamma_hat.kTable(1);
+  const auto & kvec = gamma_hat.kTable(2);
+
+  for (int freq_x = 0; freq_x < ni; ++freq_x)
+    for (int freq_y = 0; freq_y < nj; ++freq_y)
+      for (int freq_z = 0; freq_z < nk; ++freq_z)
+      {
+        if (ivec[freq_x] == 0 && jvec[freq_y] == 0 && kvec[freq_z] == 0)
+          epsilon_buffer.reciprocalSpace()[index] = _initial_strain_tensor;
+        else
+          epsilon_buffer.reciprocalSpace()[index] -=
+              gamma_hat.reciprocalSpace()[index] * stress_buffer.reciprocalSpace()[index];
```

Comment:
For this loop we could provide a helper function that takes a Lambda as an argument

```
template<typename T>
template <typename T1>
FFTData<T>::apply(T1 lambda)
{
  size_t index = 0;
  for (auto ivec : kTable(0))
    for (auto jvec : kTable(1))
      for (auto kvec : kTable(2))
        _buffer[index++] = lambda(ivec, jvec, kvec, index);
}
```

(unfortunately FFTData does not have access to the kTable yet, but we can probably store a referece to the FFTBufferBase object that owns the FFTData.)

so the call would look something like

```
epsilon_buffer.reciprocalSpace().apply([](Real ivec, Real jvec, Real kvec, std::size_t index) {
    if (ivec == 0 && jvec == 0 && kvec == 0)
      return _initial_strain_tensor;
    else
      return gamma_hat.reciprocalSpace()[index] * stress_buffer.reciprocalSpace()[index];
} );
```

Since we are using templating to pass in the lambda the compiler will inline it and should produce very fast code.

Also you ultimately will want to do

```
epsilon_buffer.reciprocalSpace().apply([](Real ivec, Real jvec, Real kvec, std::size_t index) {
    return gamma_hat.reciprocalSpace()[index] * stress_buffer.reciprocalSpace()[index];
} );
epsilon_buffer.setKZero(_initial_strain_tensor);
```

where `setKZero` needs to be implemented (and will simply write to the index that corresponds to the (0,0,0) k-vector. That way you'll write twice to that location but avoid looping over the conditional. This is likely much faster.